src/secrets/src/cache.rs

```diff
@@ -0,0 +1,740 @@
+    async fn ensure(&self, id: GlobalId, contents: &[u8]) -> Result<(), anyhow::Error> {
+        self.inner.ensure(id, contents).await?;
+
+        if self.caching_enabled() {
+            // If we reach here, then we've succeeded in ensure-ing the secret.
+            let mut write_guard = self
+                .cache
+                .write()
+                .expect("CachingSecretsController panicked!");
+            let cache_value = CacheItem::new(contents.to_vec(), Instant::now());
+
+            write_guard.insert(id, cache_value);
+        }
+
+        Ok(())
+    }
+
+    async fn delete(&self, id: GlobalId) -> Result<(), anyhow::Error> {
+        // Proactively delete the secret from the cache. If we fail to delete after this thats
+        // okay. We'll load the value back into the cache on the next read.
+        if self.caching_enabled() {
+            let mut write_guard = self
+                .cache
+                .write()
+                .expect("CachingSecretsController panicked!");
+            write_guard.remove(&id);
+        }
+
+        self.inner.delete(id).await?;
+
+        Ok(())
+    }
+
+    async fn list(&self) -> Result<Vec<GlobalId>, anyhow::Error> {
+        // No caching needed for listing secrets.
+        self.inner.list().await
+    }
+
+    fn reader(&self) -> Arc<dyn SecretsReader> {
+        let reader = CachingSecretsReader::new(
+            self.inner.reader(),
+            Arc::clone(&self.cache),
+            Arc::clone(&self.policy),
+        );
+        let reader: Arc<dyn SecretsReader> = Arc::new(reader);
+        reader
+    }
+}
+
+#[derive(Debug)]
+pub struct CachingSecretsReader {
+    /// The underlying secrets, source of truth.
+    inner: Arc<dyn SecretsReader>,
+    /// In-memory cache.
+    cache: Arc<RwLock<BTreeMap<GlobalId, CacheItem>>>,
+    /// Caching policy, can change at runtime, e.g. via LaunchDarkly.
+    policy: Arc<CachingParameters>,
+}
+
+impl CachingSecretsReader {
+    fn new(
+        reader: Arc<dyn SecretsReader>,
+        cache: Arc<RwLock<BTreeMap<GlobalId, CacheItem>>>,
+        policy: Arc<CachingParameters>,
+    ) -> Self {
+        CachingSecretsReader {
+            inner: reader,
+            cache,
+            policy,
+        }
+    }
+}
+
+#[async_trait]
+impl SecretsReader for CachingSecretsReader {
+    async fn read(&self, id: GlobalId) -> Result<Vec<u8>, anyhow::Error> {
+        // Iff our cache is enabled will we read from it.
+        if self.policy.enabled() {
+            let read_guard = self.cache.read().expect("CachingSecrets panicked!");
+            let ttl = self.policy.ttl();
+
+            // If we have a cached value we still need to check if it's expired.
+            if let Some(CacheItem { secret, ts }) = read_guard.get(&id) {
+                if Instant::now().duration_since(*ts) < ttl {
+                    return Ok(secret.clone());
+                }
```

Comment:
How load bearing is the TTL part? It seems like the cache should never be wrong since it also gets writes and deletes of secrets. With this implementation it seems that as long as the cache is enabled, a secret will always be in cache (when I think of TTL I think of expiring something out of the cache which I don't think this is doing).

Reply:
Great question! This reminded me that I forgot something.

It's not load bearing, but an effort to guard against possible inconsistencies. For example, if there is some issue, e.g. a bug or a secret changes via some back channel, then that inconsistency will only exist for at most the duration of the TTL.

I just updated the PR to check for these inconsistencies and warn us very loudly if one occurs.

Reply:
We changed the impl now, and the context of TTL did change, I think the value really is a "time to live", what do you think?

All `tracing::error`s get reported to Sentry, and I think there is some process to detect new errors? So when they fire we should learn about them pretty quickly

Reply:
Yep, the QA team monitors Sentry for new errors every day.